best/model.py

```diff
@@ -55,33 +60,17 @@ def observed_data(self, group_id: int):
     def __str__(self):
         pass
 
-    def sample(self, n_samples: int, **kwargs) -> MultiTrace:
+    def sample(self, n_samples: int, **kwargs) -> InferenceData:
         """Draw posterior samples from the model
 
         (This method is accessible primarily for internal purposes.)
         """
 
         kwargs['tune'] = kwargs.get('tune', 1000)
-        pm_major, pm_minor, *_ = pm.__version__.split('.')
-        if (int(pm_major), int(pm_minor)) < (3, 7):
-            kwargs.setdefault('nuts_kwargs', {'target_accept': 0.90})
-        else:
-            kwargs.setdefault('target_accept', 0.9)
-        max_rounds = 2
-        for r in range(max_rounds):
-            with self.model:
-                trace = pm.sample(n_samples, **kwargs)
-
-            if trace.report.ok:
-                break
-            else:
-                if r == 0:
-                    kwargs['tune'] = 2000
-                    print('\nDue to potentially incorrect estimates, rerunning sampling '
-                          'with {} tuning samples.\n'.format(kwargs['tune']), file=sys.stderr)
-                else:
-                    print('\nThe samples maybe are still not totally okay. '
-                          'Try rerunning the analysis.')
+        kwargs.setdefault('target_accept', 0.9)
+
+        with self.model:
+            trace = pm.sample(n_samples, **kwargs)
```

Comment:
This change is my one big problem with this PR. It completely drops checking for `trace.report.ok`, which sometimes is necessary. (To simulate such a case, run `best_out = best.analyze_two(group1_data, group2_data, tune=0, n_samples=1000, keep_warning_stat=True)`.)

After hours of checking your changes and debugging `InferenceData`, this is the last line that I came back to. The obvious solution is to add `return_inferencedata=False` to `pm.sample`, which would keep everything as a `MultiTrace` instead of the new `InferenceData`, so we could check `trace.report.ok`. I see no problems with that, but we'd need to drop much of our adaptations from this PR.

Alternatively, we could add `keep_warning_stat=True` and replace `trace.report.ok` with `trace.sample_stats['diverging'].any()`. This would prevent the serialisation to NetCDF, which we rely on in the tests. I don't see too many reasons to use `InferenceData` instead of `MultiTrace` (I liked this pymc-internal format more), apart from keeping these changes that you (and I) worked on in this PR – which is not an amazing reason tbh.

Do you have stronger reasons for going with `InferenceData`?

Reply:
Okay, let's go with `InferenceData`, I like that the user of BEST doesn't need to care about this PyMC-internal data structure (even though I do like it).

If you are up for making the changes, let me know until the end of the week. Thanks for your work so far.

Reply:
Hi. I only just saw this. I'm up for doing the work, but don't have time right now to think. I will look on the weekend. Thanks for putting in the hours to look at the pr!

Reply:
Hello, if you can do these changes by the end of this week, then I could review and merge them next week, otherwise it would need to wait another week while I'm AFK.

Thanks!
Laszlo

Reply:
Hey @terrycojones, any updates on this? I would fix it around Friday if you aren't (which is fine, you'd still get credited, I just don't want to drag this any further).